### ISSUE TITLE
The LeakyBucket leaks memory

### ISSUE THREAD

**Issue by vorner** [antifuchs/ratelimit_meter#17]
When I compile this rust program and run it, it's memory consumption grows steadily and quite fast. There seems to be some memory leak inside it:

```rust
extern crate ratelimit_meter;

use ratelimit_meter::{Decider, LeakyBucket};

fn main() {
    let mut bucket = LeakyBucket::per_second(1_000_000).unwrap();
    loop {
        drop(bucket.check());
    }
}
```

I'll have a look if I see something obvious in the code.

**Comment by antifuchs** [comment]
Eeep, this sounds super weird! I'll take a look, suspect it's me mishandling the `Arc`!

**Comment by antifuchs** [comment]
Sooo, yep, I have reproduced this with both the `LeakyBucket` and `GCRA`. I am pretty sure it's related to the way I use crossbeam_epoch - I misunderstood it as a way to get realistic atomic behavior without the `unsafe`, but it's really an atomic pointer implementation hooked up to a garbage collector, and I'm likely doing the garbage collection parts wrong.

I'll investigate other things, but we may just have to switch back to Mutexes (and pollute the error returns again, counteracting #12), oh no.

**Comment by vorner** [comment]
I think two things to point out:

* Mutexes are often faster ‒ especially if there's very little contention on them. Atomic algorithms might have an edge at some cases, but it's usually when the scenario is read-heavy and contended.
* Mutex fails to lock only in cases when someone panicked inside a lock. It's quite common to either ignore such errors (if a panic in critical section can't leave any mess or if you are sure you can't panic there), or unwrap and propagate the panic. There are some other mutexes (in parking lot, for example) that don't implement panic poisoning at all.

**Comment by antifuchs** [comment]
Yeah, I looking at the old benchmarks (where I'd been using Mutexes), it wasn't that slow (the crossbeam implementation actually slowed it down quite a bit, in favor of multithreaded perf). I'll rewrite it to be threadsafe-by-default & use a mutex. Pretty sure we can come up with a way to talk about poisoning, too.

**Comment by antifuchs** [comment]
Fixed in v2.0.0, released just now!
